### DISCUSSION TITLE
bloblang syntax about flatten arrays in array and set a dynamic key in an object

### DISCUSSION THREAD

**Discussion by fearfate** [redpanda-data/connect#3068] [Q&A] [answered]
```
root = this.flatten()
# In: [["MIT", "GPL"], ["APACHE"]]
# Out: ["MIT", "GPL", "APACHE"] 
```

```
root."key:%v".format(123456) = "value"
# In: {"key1": "value1"}
# Out: {"key1": "value1", "key:123456": "value"}
```

**Comment by mihaitodor** [comment, marked as answer]
Hey @fearfate you can't assign to expressions like that. Maybe try this:

```coffee
root = this.merge({("key:%v".format(123456)): "value"})
```

The extra `()` are necessary, unfortunately.

Regarding `flatten()`, I'm not sure what the question is.

**Reply by fearfate** [reply]
I want flatten two-dimensional array into one-dimensional array, example as input `[["MIT", "GPL"], ["APACHE"]]` flatten to `["MIT", "GPL", "APACHE"]`, but the `flatten()` method outputs `[["MIT", "GPL"], ["APACHE"]]`

**Reply by mihaitodor** [reply]
Looks right to me in `redpanda-connect blobl server`:
<img width="558" alt="image" src="https://github.com/user-attachments/assets/c8e2ce0b-8722-4925-8e91-6fe2d8a78119">

**Reply by fearfate** [reply]
I customized a methods:

```go
import (
	"github.com/github/go-spdx/v2/spdxexp"
)

func init() {
	extraLicenseSpec := bloblang.NewPluginSpec().
		Description("extra license expression to license ids")
	if err := bloblang.RegisterMethodV2("extract_license",
		extraLicenseSpec,
		func(args *bloblang.ParsedParams) (bloblang.Method, error) {
			return bloblang.StringMethod(ExtractLicense), nil
		}); err != nil {
		panic(err)
	}
	extraLicenseSpec = extraLicenseSpec.
		Param(bloblang.NewStringParam("license_expression"))
	if err := bloblang.RegisterFunctionV2("extract_license",
		extraLicenseSpec,
		func(args *bloblang.ParsedParams) (bloblang.Function, error) {
			purl, err := args.GetString("purl")
			if err != nil {
				return nil, err
			}

			return func() (any, error) {
				return ExtractLicense(purl)
			}, nil
		}); err != nil {
		panic(err)
	}
}

func ExtractLicense(s string) (any, error) {
	return spdxexp.ExtractLicenses(s)
}
```

input:

```
{"name": ["MIT AND GPL-2.0", "APACHE-2.0"]}
```

bloblang:

```
root = this.name.map_each(e -> e.extract_license()).flatten()
```

![image](https://github.com/user-attachments/assets/dca8db32-c542-43bb-a89b-494d6b9f0671)

**Reply by mihaitodor** [reply]
The issue is that [`ExtractLicenses()`](https://github.com/github/go-spdx/blob/acea5bc11f616613ef809d718dde95aa6ea7e80f/spdxexp/extracts.go#L5) returns `[]string`, but you must return a scalar Go type that is recognised by the Bloblang `type()` method, or either `map[string]any` or `[]any`. In your case, it should be `[]any` and you need to convert your `[]string` slice to that. I see the Bloblang API docs don't mention this, so it might be worth raising a documentation bug on https://github.com/redpanda-data/benthos.

**Reply by fearfate** [reply]
It works! I appreciate your help.
